VirtoCommerce.CustomerModule.Data/Handlers/LogChangesMemberChangedEventHandler.cs

```diff
@@ -0,0 +1,249 @@
+                        {
+                            AddAddressRemovedOperation(result, member, target);
+                        }
+                    });
+        }
+
+        protected virtual IEnumerable<string> GetListChanges<T>(
+            IEnumerable<T> originalCollection,
+            IEnumerable<T> modifiedCollection,
+            Func<List<string>, Action<EntryState, T, T>> compareAction)
+        {
+            if (originalCollection == null || modifiedCollection == null)
+            {
+                return Enumerable.Empty<string>();
+            }
+
+            var result = new List<string>();
+            modifiedCollection
+                .Where(x => x != null).ToList()
+                .CompareTo(
+                    originalCollection.Where(x => x != null).ToList(),
+                    EqualityComparer<T>.Default,
+                    compareAction(result));
+            return result;
+        }
+
+        private static void AddMemberPropertyChangedOperation(
+            List<string> result, GenericChangedEntry<Member> changedEntry, Difference difference)
+        {
+            result.Add(
+                string.Format(
+                    MemberResources.MemberPropertyChanged,
+                    changedEntry.OldEntry.MemberType,
+                    changedEntry.NewEntry.Name,
+                    difference.Name,
+                    difference.OldValue,
+                    difference.NewValue));
+        }
+
+        private void AddPhoneRemovedOperation(List<string> result, Member member, string phone)
+        {
+            result.Add(string.Format(MemberResources.PhoneRemoved, member.MemberType, member.Name, phone));
+        }
+
+        private void AddPhoneAddedOperation(List<string> result, Member member, string phone)
+        {
+            result.Add(string.Format(MemberResources.PhoneAdded, member.MemberType, member.Name, phone));
+        }
+
+        private void AddEmailRemovedOperation(List<string> result, Member member, string email)
+        {
+            result.Add(string.Format(MemberResources.EmailRemoved, member.MemberType, member.Name, email));
+        }
+
+        private void AddEmailAddedOperation(List<string> result, Member member, string email)
+        {
+            result.Add(string.Format(MemberResources.EmailAdded, member.MemberType, member.Name, email));
+        }
+
+        private void AddAddressRemovedOperation(List<string> result, Member member, Address address)
+        {
+            result.Add(
+                string.Format(MemberResources.AddressRemoved, member.MemberType, member.Name, StringifyAddress(address)));
+        }
+
+        private void AddAddressAddedOperation(List<string> result, Member member, Address address)
+        {
+            result.Add(
+                string.Format(MemberResources.AddressAdded, member.MemberType, member.Name, StringifyAddress(address)));
+        }
+
+        private static string StringifyAddress(Address address)
```

Comment:
The Address type has a overridden ToString method you might use it without the self implementation

Reply:
Isn't it risky to rely on `ToString()`? I would use it with please, just thought it's safer.